src/constructors.jl

```diff
@@ -0,0 +1,147 @@
+abstract type AbstractHMCSampler <: AbstractMCMC.AbstractSampler end
+
+##############
+### Custom ###
+##############
+
+"""
+    HMCSampler
+
+A `AbstractMCMC.AbstractSampler` for kernels in AdvancedHMC.jl.
+
+# Fields
+
+$(FIELDS)
+
+# Notes
+
+Note that all the fields have the prefix `initial_` to indicate
+that these will not necessarily correspond to the `kernel`, `metric`,
+and `adaptor` after sampling.
+
+To access the updated fields use the resulting [`HMCState`](@ref).
+"""
+Base.@kwdef struct HMCSampler{
+    K<:AbstractMCMCKernel,
+    M<:AbstractMetric,
+    A<:Adaptation.AbstractAdaptor,
+} <: AbstractHMCSampler
+    "[`AbstractMCMCKernel`](@ref)."
+    kernel::K
+    "[`AbstractMetric`](@ref)."
+    metric::M
+    "[`AbstractAdaptor`](@ref)."
+    adaptor::A
+    "Adaptation steps if any"
+    n_adapts::Int = 0
+end
+
+############
+### NUTS ###
+############
+"""
+    NUTS(n_adapts::Int, δ::Float64; max_depth::Int=10, Δ_max::Float64=1000.0, init_ϵ::Float64=0.0)
+
+No-U-Turn Sampler (NUTS) sampler.
+
+Usage:
+
+```julia
+NUTS()            # Use default NUTS configuration.
+NUTS(1000, 0.65)  # Use 1000 adaption steps, and target accept ratio 0.65.
+```
+
+Arguments:
+
+- `n_adapts::Int` : The number of samples to use with adaptation.
+- `δ::Float64` : Target acceptance rate for dual averaging.
+- `max_depth::Int` : Maximum doubling tree depth.
+- `Δ_max::Float64` : Maximum divergence during doubling tree.
+- `init_ϵ::Float64` : Initial step size; 0 means automatically searching using a heuristic procedure.
+
+"""
+Base.@kwdef struct NUTS <: AbstractHMCSampler
+    n_adapts::Int                      # number of samples with adaption for ϵ
+    δ::Float64                         # target accept rate
```

Comment:
This is too strict. It should be possible to use, say, `Float32`. Similar with all the other fields restricted to `Float64`.